src/vs/workbench/contrib/chat/browser/chatContentParts/toolInvocationParts/chatTerminalToolProgressPart.ts

```diff
@@ -106,42 +142,325 @@ export class ChatTerminalToolProgressPart extends BaseChatToolInvocationSubPart
+				inlineTerminal.style.setProperty('background-color', theme.background || 'transparent');
+				inlineTerminal.style.setProperty('color', theme.foreground || 'inherit');
+			}
+		}
+
+		this._outputBody.replaceChildren(content);
+		this._outputContent = content;
+		if (!this._outputScrollbar) {
+			this._outputScrollbar = this._register(new DomScrollableElement(this._outputBody, {
+				vertical: ScrollbarVisibility.Auto,
+				horizontal: ScrollbarVisibility.Auto,
+				handleMouseWheel: true
+			}));
+			const scrollableDomNode = this._outputScrollbar.getDomNode();
+			scrollableDomNode.tabIndex = 0;
+			scrollableDomNode.style.maxHeight = `${MAX_TERMINAL_OUTPUT_PREVIEW_HEIGHT}px`;
+			this._outputContainer.appendChild(scrollableDomNode);
+			this._ensureOutputResizeObserver();
+			this._outputContent = undefined;
+		} else {
+			this._ensureOutputResizeObserver();
+		}
+
+		return true;
+	}
+
+	private _scrollOutputToBottom(): void {
+		if (!this._outputScrollbar) {
+			return;
+		}
+		const dimensions = this._outputScrollbar.getScrollDimensions();
+		this._outputScrollbar.setScrollPosition({ scrollTop: dimensions.scrollHeight });
+	}
+
+	private _scheduleOutputRelayout(): void {
+		dom.getActiveWindow().requestAnimationFrame(() => {
+			this._layoutOutput();
+			this._scrollOutputToBottom();
+		});
+	}
+
+	private _layoutOutput(): void {
+		if (!this._outputScrollbar || !this._outputContainer.classList.contains('expanded')) {
+			return;
+		}
+		const scrollableDomNode = this._outputScrollbar.getDomNode();
+		const viewportHeight = Math.min(this._outputBody.scrollHeight, MAX_TERMINAL_OUTPUT_PREVIEW_HEIGHT);
+		scrollableDomNode.style.height = `${viewportHeight}px`;
+		this._outputScrollbar.scanDomNode();
+	}
+
+	private _ensureOutputResizeObserver(): void {
+		if (this._outputResizeObserver || !this._outputScrollbar) {
+			return;
+		}
+		const observer = new ResizeObserver(() => this._layoutOutput());
+		observer.observe(this._outputContainer);
+		this._outputResizeObserver = observer;
+		this._register(toDisposable(() => {
+			observer.disconnect();
+			this._outputResizeObserver = undefined;
+		}));
+	}
+
+	private async _collectOutput(terminalInstance: ITerminalInstance | undefined): Promise<{ text: string; truncated: boolean }> {
+		const storedOutput = this._terminalData.output;
+		if (storedOutput?.html) {
+			return { text: storedOutput.html, truncated: storedOutput.truncated ?? false };
+		}
+		if (!terminalInstance) {
+			return { text: '', truncated: false };
+		}
+		const xterm = await terminalInstance.xtermReadyPromise;
+		if (!xterm) {
+			return { text: '', truncated: false };
+		}
+		const command = this._resolveCommand(terminalInstance);
+		if (!command?.endMarker) {
+			return { text: '', truncated: false };
+		}
+		const text = await xterm.getCommandOutputAsHtml(command, CHAT_TERMINAL_OUTPUT_MAX_PREVIEW_LINES);
+		if (!text) {
+			return { text: '', truncated: false };
+		}
+
+		return { text, truncated: false };
```

Comment:
This `if (!text)` seems redundant as `getCommandOutputAsHtml` returns a string and when it's `''` the behavior would be the same if it fell through.